File: content/problems/0501-find-mode-in-binary-search-tree/solutions/inorder-run-count/solution.cpp
class Solution {
    optional<int> previous;
    int count=0,best=0;
    vector<int> modes;
    void inorder(TreeNode* node){
        if(!node)return;
        inorder(node->left);
        count=previous&&*previous==node->val?count+1:1;
        if(count>best){
            best=count;
            modes.clear();
            modes.push_back(node->val);
        }else if(count==best)modes.push_back(node->val);
        previous=node->val;
        inorder(node->right);
    }
    public:vector<int> findMode(TreeNode* root){
        inorder(root);
        return modes;
    }
};

File: content/problems/0501-find-mode-in-binary-search-tree/solutions/inorder-run-count/Solution.kt
class Solution {
    private var previous:Int?=null;
    private var count=0;
    private var best=0;
    private val modes=mutableListOf<Int>();
    fun findMode(root:TreeNode?):IntArray{
        fun inorder(node:TreeNode?){
            if(node==null)return;
            inorder(node.left);
            count=if(previous==node.`val`)count+1 else 1;
            if(count>best){
                best=count;
                modes.clear();
                modes.add(node.`val`)
            }else if(count==best)modes.add(node.`val`);
            previous=node.`val`;
            inorder(node.right)
        };
        inorder(root);
        return modes.toIntArray()
    }
}
